be/src/exec/pipeline/exchange/mem_limited_chunk_queue.h

```diff
@@ -0,0 +1,244 @@
+                    return true;
+                }
+                return false;
+            }
+            // still has un-consumed data in current block
+            return true;
+        }
+
+        // return an copy of next cell
+        Iterator next() const {
+            DCHECK(has_next());
+            Iterator iter(block, index);
+            iter.move_to_next();
+            return iter;
+        }
+        void move_to_next() {
+            DCHECK(has_next());
+            index++;
+            DCHECK_LE(index, block->cells.size());
+            if (index == block->cells.size()) {
+                block = block->next;
+                index = 0;
+            }
+        }
+
+        Cell* get_cell() {
+            DCHECK(block != nullptr);
+            DCHECK(index < block->cells.size()) << ", invalid index: " << index << ", size: " << block->cells.size();
+            return &(block->cells[index]);
+        }
+
+        Block* get_block() { return block; }
+    };
+
+    static const size_t kDefaultBlockSize = 8L * 1024 * 1024;
+    static const size_t kDefaultMaxUnconsumedBytes = 16L * 1024 * 1024;
+    static const size_t kDefaultMemoryLimit = 16L * 1024 * 1024;
+
+public:
+    struct Options {
+        int32_t plan_node_id = 0;
+        int encode_level = 7;
+        // memory block size
+        size_t block_size = kDefaultBlockSize;
+        size_t memory_limit = kDefaultMemoryLimit;
+        size_t max_unconsumed_bytes = kDefaultMaxUnconsumedBytes;
+        spill::BlockManager* block_manager = nullptr;
+        workgroup::WorkGroupPtr wg;
+    };
+
+    MemLimitedChunkQueue(RuntimeState* state, int32_t consumer_number, Options opts);
+    ~MemLimitedChunkQueue();
+
+    Status init_metrics(RuntimeProfile* parent);
+
+    Status push(const ChunkPtr& chunk);
+    bool can_push();
+
+    StatusOr<ChunkPtr> pop(int32_t consumer_index);
+    bool can_pop(int32_t consumer_index);
+
+    void open_consumer(int32_t consumer_index);
+
+    void close_consumer(int32_t consumer_index);
+
+    void open_producer();
+
+    void close_producer();
+
+    void enter_release_memory_mode();
+
+private:
+    void _update_progress(Iterator* iter = nullptr);
+
+    void _close_consumer(int32_t consumer_index);
+
+    Status _flush();
+    // trigger flush task
+    Status _submit_flush_task();
+    // reload block from disk
+    Status _load(Block* block);
+
+    Status _submit_load_task(Block* block);
+
+    void _evict_loaded_block();
+
+    inline void _update_io_task_status(const Status& status) {
+        if (status.ok() || _io_task_status.load() != nullptr) {
+            return;
+        }
+        Status* old_status = nullptr;
+        if (_io_task_status.compare_exchange_strong(old_status, &_status)) {
+            _status = status;
+        }
+    }
+    Status _get_io_task_status() const {
+        auto* status = _io_task_status.load();
+        return status == nullptr ? Status::OK() : *status;
+    }
+
+#ifdef BE_TEST
```

Comment:
we don't have to add this. UT compile with flag -fno-access-control

Reply:
fixed